### PR TITLE
linux service changes using nodejs

### PR DESCRIPTION
**Pull request by kasubram**
This is only the service side code changes. Its not integrated with agent command line yet.
Next PR would contain Agent.Listener changes to hook this service into systemd.

### CODE REVIEW

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs [hunk 1]
```diff
@@ -1,14 +1,281 @@
...
+        public override void ConfigureService(
+            AgentSettings settings,
+            Dictionary<string, string> args,
+            bool enforceSupplied)
+        {
+            Trace.Info(nameof(ConfigureService));
+
+            CalculateServiceName(settings, ServiceNamePattern, ServiceDisplayNamePattern);
+
+            if (!CheckIfSystemdExists())
```
Comment (bryanmacfarlane) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: Are we starting with just systemd?  Is it possible to install/apt-get/whatever and get systemd on ubuntu 14.04?  The reason I ask is there's currently too many issues with core CLR on ubuntu 15.10 (I finally gave up recently - just not ready).  I expect that to get resolved before they ship but I'm wondering how we evaluate this work in the short term.  Cent OS?

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs [hunk 2]
```diff
@@ -1,14 +1,281 @@
...
+                    chown(file.Key, stat.pw_uid, stat.pw_gid);
+
+                    Trace.Info("Changing permission of {0} to {1}", file.Key, file.Value);
+                    chmod(file.Key, file.Value);
+                }
+            }
+            catch (Exception ex)
+            {
+                Trace.Error(ex);
+                _term.WriteError(StringUtil.Loc("CannotChangeOwnership"));
```
Comment (bryanmacfarlane) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: It's not safe to just eat this exception right?  Doesn't it leave it in a broken state?  I assume this is important right?

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs [hunk 3]
```diff
@@ -1,14 +1,281 @@
...
+        private void EnableService(string serviceName)
+        {
+            Trace.Info(nameof(EnableService));
+            try
+            {
+                ExecuteSystemdCommand("enable " + serviceName);
+            }
+            catch (Exception)
+            {
+                _term.WriteError(StringUtil.Loc("LinuxServiceStartFailed"));
```
Comment (bryanmacfarlane) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: It's not safe to just eat this exception right?  Doesn't it leave it in a broken state?  I assume this is important right?  This applies to all these private functions

File: src/Agent.Listener/Configuration/WindowsServiceControlManager.cs [hunk 4]
```diff
@@ -51,25 +51,16 @@ public override Task ConfigureServiceAsync(AgentSettings settings, Dictionary<st
...
             // TODO check and unconfigure/configure the service using advapi32.dll service install methods using pinvoke
-
-            return Task.FromResult(0);
         }
 
-        public bool CheckServiceExists(string serviceName)
+        public override void StartService(string serviceName)
         {
-            return false;
+            throw new NotImplementedException();
```
Comment (bryanmacfarlane) [inline src/Agent.Listener/Configuration/WindowsServiceControlManager.cs]: cool - I'll implement the details here

File: src/Agent.Service/Linux/AgentService.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	    listener.stderr.on('data', (data) => {
+		trace.error(data);
+	    });
+
+	    listener.on('close', (code, signal) => {
+		trace.info(`Agent listenere exited with error code ${code}`);
+
+		if (code == 1)
+		{
+		    trace.error('Can not start agent listener process, check logs for error detail');
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/AgentService.ts]: We may want to console.log here as well.  I know that at least on OSX all stdout and stderr out goes to the service (console) log.

File: src/Agent.Service/Linux/AgentService.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env node
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+/// <reference path="./definitions/node.d.ts"/>
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/AgentService.ts]: I would not compile this.  There's not enough code here (89 lines) to justify pulling in typescript, gulp dev dependencies etc...

File: src/Agent.Service/Linux/definitions/node.d.ts [hunk 7]
```diff
@@ -0,0 +1,1329 @@
+// Type definitions for Node.js v0.10.1
+// Project: http://nodejs.org/
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/definitions/node.d.ts]: You don't need this - see my other comment about not compiling this one script

File: src/Agent.Service/Linux/gulpfile.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+var fs = require('fs');
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/gulpfile.js]: don't need this

File: src/Agent.Service/Linux/package.json [hunk 9]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/package.json]: don't need this

File: src/Agent.Service/Linux/trace.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (bryanmacfarlane) [inline src/Agent.Service/Linux/trace.ts]: I don't think we need a trace do we?  See my question about console out.  I believe that's the story for service feedback on both linux and osx

File: src/dev.sh [hunk 11]
```diff
@@ -53,9 +54,39 @@ function rundotnet ()
     done   
 }
 
+function buildagentservice()
+{
+    if [[ "$define_os" == 'OS_LINUX' ]]; then
+	echo Building Linux Agent Service ...
+
+	#TODO Assume node/gulp already installed?
```
Comment (bryanmacfarlane) [inline src/dev.sh]: shouldn't need.  Make sure it's part of the Misc files to copy to layout

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs (hunk 1 above)
Comment (kasubram) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: Yeah, I tried on 15.10 as well, it didn't work either. I tried upgrading 14.04 to use systemd which didn't work either. It requires kernal, grub change.. after that I started to get kernel panic. I couldn't recover. I am right now using CentOS7.1 where everything works good.

File: src/Agent.Listener/Configuration/WindowsServiceControlManager.cs (hunk 4 above)
Comment (kasubram) [inline src/Agent.Listener/Configuration/WindowsServiceControlManager.cs]: I am working on this.. Did you mean for OSX?

File: src/Agent.Service/Linux/AgentService.ts (hunk 6 above)
Comment (kasubram) [inline src/Agent.Service/Linux/AgentService.ts]: ok. :) Just keep it as .js file?

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs (hunk 2 above)
Comment (kasubram) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: correct. if something goes wrong service can't start. However there is nothing much we can do other than reporting it and this is not retryable. It makes sense to throw error..

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs (hunk 3 above)
Comment (kasubram) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: I was thinking not able to start the service can be a warning though.. changed it throw error anyway..

File: src/Agent.Service/Linux/AgentService.ts (hunk 5 above)
Comment (kasubram) [inline src/Agent.Service/Linux/AgentService.ts]: Done

File: src/Agent.Service/Linux/definitions/node.d.ts (hunk 7 above)
Comment (kasubram) [inline src/Agent.Service/Linux/definitions/node.d.ts]: Done

File: src/Agent.Service/Linux/gulpfile.js (hunk 8 above)
Comment (kasubram) [inline src/Agent.Service/Linux/gulpfile.js]: Done :)

File: src/Agent.Service/Linux/package.json (hunk 9 above)
Comment (kasubram) [inline src/Agent.Service/Linux/package.json]: Done :) :)

File: src/Agent.Service/Linux/trace.ts (hunk 10 above)
Comment (kasubram) [inline src/Agent.Service/Linux/trace.ts]: Done :) :) :) 
Yes, we could do journalctl -u service.name and see all the logs from the service

File: src/dev.sh (hunk 11 above)
Comment (kasubram) [inline src/dev.sh]: Done

**Comment by bryanmacfarlane** [comment]
OK - looks good enough for now.  I may have more thoughts after I consume the changes and start running.

File: src/Agent.Listener/Configuration/LinuxServiceControlManager.cs [hunk 12]
```diff
@@ -1,14 +1,289 @@
...
+
+using Microsoft.VisualStudio.Services.Agent.Util;
 
 namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
 {
     public class LinuxServiceControlManager : ServiceControlManager
     {
-        public override Task ConfigureServiceAsync(AgentSettings settings, Dictionary<string, string> args, bool enforceSupplied)
+        // This is the name you would see when you do `systemctl list-units | grep vsts`
+        private const string ServiceNamePattern = "vsts.agent.{0}.{1}.service";
```
Comment (GitHubSriramB) [inline src/Agent.Listener/Configuration/LinuxServiceControlManager.cs]: these can also move into the base class?

File: src/Misc/layoutbin/AgentService.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            listener.stdout.on('data', (data) => {
+                console.log(data);
+            });
+
+            listener.stderr.on('data', (data) => {
+                console.log(data);
+            });
+
+            listener.on('close', (code) => {
+                console.log(`Agent listenere exited with error code ${code}`);
```
Comment (GitHubSriramB) [inline src/Misc/layoutbin/AgentService.js]: fix typo

File: src/Test/L0/Listener/Configuration/LinuxServiceControlManagerL0.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            tc.SetSingleton<IProcessInvoker>(_processInvoker.Object);
+            return tc;
+        }
+
+#if OS_LINUX
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Common")]
+#endif
+        public void LinuxServiceControlManagerShouldConfigureCorrectly()
```
Comment (GitHubSriramB) [inline src/Test/L0/Listener/Configuration/LinuxServiceControlManagerL0.cs]: can we verify the chmod calls happening as expected as well?

**Pull request merged**